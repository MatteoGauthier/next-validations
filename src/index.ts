export * from './validation';
export * from './withValidation';
