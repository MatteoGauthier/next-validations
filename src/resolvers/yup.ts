
import * as Yup from 'yup';
import type Lazy from 'yup/lib/Lazy';

export function yupResolver<T extends Yup.AnyObjectSchema | Lazy<any>>(schema:T) {
    return {
        validate: (data: unknown) => schema.validateSync(data),
    }
}
