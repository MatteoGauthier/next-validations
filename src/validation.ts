import { yupResolver } from './resolvers/yup';
import { fastestValidatorResolver } from './resolvers/fastest-validator';

export type SCHEMA_TYPE =
  | 'Yup'
  | 'FastestValidator'
  | 'Zod' // TODO add Zod validation
  | 'SuperStruct' // TODO SuperStruct SuperStruct validation
  | 'Joi' // TODO SuperStruct SuperStruct validation
  | 'Vest' // TODO SuperStruct Zod validation
  | 'ClassValidator' // TODO SuperStruct ClassValidator validation
  | 'IoTS' // TODO SuperStruct IoTS validation
  | 'Nope'; // TODO SuperStruct IoTS validation

export function createResolver(type: SCHEMA_TYPE, schema: any) {
  switch (type) {
    case 'Yup':
      return yupResolver(schema);
    case 'FastestValidator':
      return fastestValidatorResolver(schema);

    default:
      throw new Error(`Does not support ${type} validation yet!`);
  }
}
