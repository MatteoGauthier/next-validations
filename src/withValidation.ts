import { NextApiRequest, NextApiResponse } from 'next';

import { SCHEMA_TYPE, createResolver } from './validation';

type NextHandler = (err?: Error) => void;

type ValidationHoF = {
  type: SCHEMA_TYPE;
  mode?: 'body' | 'query' | 'headers';
  schema: unknown;
};

export function withValidation({
  type,
  schema,
  mode = 'query',
}: ValidationHoF) {
  return (
    handler?: (req: NextApiRequest, res: NextApiResponse<any>) => any
  ) => {
    return async (
      req: NextApiRequest,
      res: NextApiResponse,
      next?: NextHandler
    ) => {
      try {
        const resolver = createResolver(type, schema);

        resolver.validate(req[mode]);

        if (!!next) {
          return next();
        }

        if (handler) return handler(req, res);

        res.status(404).end();
      } catch (error) {
        process.env.API_VALIDATION ? res.status(400).send(error) : res.status(404).end()
      }
    };
  };
}
