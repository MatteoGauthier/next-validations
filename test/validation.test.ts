import * as yup from 'yup';
import { createResolver } from '../src/validation';

describe('Yup Validation', () => {
  it('should create yup resolve base on schema', () => {
    const resolver = createResolver('Yup', {});
    expect(resolver).toMatchSnapshot();
  });

  it('should return true', () => {
    const schema = yup.object().shape({
      name: yup.string().required(),
      age: yup
        .number()
        .required()
        .positive()
        .integer(),
      email: yup.string().email(),
      website: yup.string().url(),
      createdOn: yup.date().default(function() {
        return new Date();
      }),
    });
    const resolver = createResolver('Yup', schema);
    const isValid = resolver.validate({
      name: 'jimmy',
      age: 24,
    });
    expect(isValid).toBeTruthy();
  });

  it('should return false', () => {
    let schema = yup.object().shape({
      name: yup.string().required(),
      age: yup
        .number()
        .required()
        .positive()
        .integer(),
      email: yup.string().email(),
      website: yup.string().url(),
      createdOn: yup.date().default(function() {
        return new Date();
      }),
    });
    const resolver = createResolver('Yup', schema);
    try {
      resolver.validate({
        name: 'jimmy',
      });
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });
});

describe('Yoi Validation', () => {
  it('should create yoi resolve base on schema', () => {
    try {
      createResolver('Joi', {});
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });
});
