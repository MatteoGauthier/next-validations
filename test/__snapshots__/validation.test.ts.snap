// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FastestValidator Validation should create resolve base on schema 1`] = `
Object {
  "validate": [Function],
}
`;

exports[`FastestValidator Validation should return false 1`] = `
Array [
  Object {
    "actual": undefined,
    "field": "status",
    "message": "The 'status' field is required.",
    "type": "required",
  },
]
`;

exports[`Yoi Validation should return false 1`] = `[ValidationError: "email" is required]`;

exports[`Yup Validation should create yup resolve base on schema 1`] = `
Object {
  "validate": [Function],
}
`;

exports[`Yup Validation should return false 1`] = `[ValidationError: age is a required field]`;

exports[`Zod Validation should return false 1`] = `
[ZodError: [
  {
    "code": "too_small",
    "minimum": 8,
    "type": "string",
    "inclusive": true,
    "message": "Should be at least 8 characters",
    "path": [
      "username"
    ]
  }
]]
`;
